tests/ConformanceTestOnSamples-report.txt

```diff
@@ -70,11 +70,11 @@ FAIL: CaptureConvertedUnspecToOther.java: no unexpected facts
 FAIL: CaptureConvertedUnspecToOtherUnionNull.java: no unexpected facts
 FAIL: CaptureConvertedUnspecToOtherUnspec.java: no unexpected facts
 PASS: CastOfCaptureOfNotNullMarkedUnboundedWildcardForObjectBoundedTypeParameter.java: no unexpected facts
-PASS: CastOfCaptureOfUnboundedWildcardForNotNullMarkedObjectBoundedTypeParameter.java: no unexpected facts
+FAIL: CastOfCaptureOfUnboundedWildcardForNotNullMarkedObjectBoundedTypeParameter.java: no unexpected facts
```

Comment:
It is correct for the checker to produce an error here. The reason that it's showing up as `FAIL` is that the conformance-test framework doesn't support `jspecify_nullness_not_enough_information` yet, [only `jspecify_nullness_mismatch`](https://github.com/jspecify/jspecify-reference-checker/blob/75aa1de2a8f88f8c23347d86465844df9887874c/conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java#L45).

(`Object*` is not a subtype of `Object*` so as to indicate that, if you don't know what you have and you don't know what you need, we can't prove that it's safe. Our expectation is that few checkers (if any) will produce even a warning in this case—or even in a more-likely-to-be-unsafe case in which you have an `Object?` and need an `Object*`. Still, the tests label such cases accordingly. This labeling may help tools like the Checker Framework that make specific assumptions about unannotated code: If a line is labeled with `jspecify_nullness_mismatch`, a tool like the Checker Framework can consider itself to "pass" whether it produces an error there or not. This is in contrast to cases in which it should always produce an error and cases in which it never should.)

Reply:
Thanks, so I did remember correctly that that is how unspecified nullness is supposed to work.

I've opened https://github.com/jspecify/jspecify/pull/487 to add the optional error.

(In your penultimate sentence: `If a line is labeled with jspecify_nullness_mismatch, a tool like the Checker Framework can consider itself to "pass" whether it produces an error there or not.` Here you meant `jspecify_nullness_not_enough_information`, right?)

Reply:
Oops, yes, sorry.

Reply:
I'm sorry, I was not reading carefully enough :( I had been looking at [line 29](https://github.com/jspecify/jspecify/blob/ec998d519a1716e034a4b1a538d42e9e7ed5c3a6/samples/CastOfCaptureOfUnboundedWildcardForNotNullMarkedObjectBoundedTypeParameter.java#L29), where we do want an error in strict mode.

I agree that line 26 should _not_ be an error even in strict mode.

I will see if I can dig up some commits related to this error, but off the top of my head:

IIRC, it seemed weird to me that the checker was checking whether the captured type was a subtype of the type-parameter bound. I don't recall a similar check in the JLS, perhaps because such a check is necessarily true given how the captured type is constructed :) I wondered if the checker behavior was a holdover from before the checker had an implementation of capture conversion.

In fact, I thought I had some memory that the checker was actually checking whether _the wildcard bound_ was a subtype of the type-parameter bound, which seemed even weirder: Java will let you write `class Foo<T extends CharSequence>` and still write `Foo<? extends Object>`, so why wouldn't the checker let you write `class Foo<T extends Object>` and still write `Foo<? extends @Nullable Object>`? That behavior doesn't seem to be what we're seeing here, since we see the type `capture#576 of ? extends Object*`, which clearly is taking the type-parameter bound into account. So either I'm remembering wrong, or I put in hacks that changed the behavior, or something changed in the checker.

Reply:
I am still not sure where the code that makes the main branch avoid this error is :\

I thought maybe https://github.com/jspecify/jspecify-reference-checker/commit/89d9b720a64f1b0bb2e6d5add574321f93339625#diff-ce82a59b06d0d82bfc6d06835d1b06ac15619bc3a71617d32189d041f02ca9c0R270, but then I found that I'd removed that cocde in https://github.com/jspecify/jspecify-reference-checker/commit/8ba39b04980f3739eb9c4a0351ee8931683c99a0

I thought maybe https://github.com/jspecify/checker-framework/commit/00cc0561ab759922eba86ff6980787b848cf907a, but then I found that I'd removed that code in https://github.com/jspecify/checker-framework/commit/72c5c16a026fc42b2b076db517b03cab20ed97a4.

Maybe https://github.com/jspecify/jspecify-reference-checker/commit/9348e76115088149d55d9495a5c720a5f1d6cd34#diff-ce82a59b06d0d82bfc6d06835d1b06ac15619bc3a71617d32189d041f02ca9c0R894 is somehow related?

I also have open https://github.com/jspecify/jspecify-reference-checker/commit/74f564f415046f6fc4e0200a7990a0db603db9c8, https://github.com/jspecify/jspecify-reference-checker/commit/7e662e31e4553d81bf4d38cb197b21f54c0fcaca, and https://github.com/jspecify/jspecify-reference-checker/commit/824113a212a7ea001e6557fb800a5d73e8f18232, but I'm not convinced that any of that is relevant.

Reply:
OK, I started just diffing `checker-framework` against upstream, and the code jumped right out to me. At least, I _assume_ this is the right code:

https://github.com/jspecify/checker-framework/blob/b84882d9b2f77c99081118cc7ad095dfcc3484f5/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeValidator.java#L562

(That code was a small part of the large merge commit https://github.com/jspecify/checker-framework/commit/2e588e78bd81ee52151bda45720120f3cd83c52a.)

Reply:
Thanks a lot for tracking this down!
I'll look into fixing this in the reference checker or the framework.
Besides the error here, these incorrect captures are also the cause for #164.

Reply:
Ah, do you think that it's possible that #164 is related to https://github.com/jspecify/checker-framework/blob/b84882d9b2f77c99081118cc7ad095dfcc3484f5/framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java#L5017 (https://github.com/jspecify/checker-framework/issues/47)?

That is, could there be two issues here?
1. The upstream checker performs a check that the wildcard bound (or capture thereof?) is in bounds (of the type-parameter bound? of the wildcard bound?). That leads to failures for JSpecify because `Object*` is not a subtype of `Object*`. My "fix" for that was to remove the possibly unnecessary check entirely. (cause of https://github.com/jspecify/jspecify/pull/487)?)
2. When we do perform capture conversion, we get it wrong. (cause of #164?)

The reason that it seems strange for #164 to be related to this issue is that #164 doesn't involve unspecified nullness.

But I guess that, if the tests of #164 start passing with this change, then it's a fix :) Still, maybe it's more that we're _hiding_ the failure (by no longer checking the incorrect capture against anything) than fixing it (by making the results of capture correct)? I don't know :\

I should also write up something about type parameters in general that I've been thinking about recently. It might end up saving us some work... maybe....

Reply:
https://github.com/jspecify/jspecify-reference-checker/pull/165/commits/9ee2796e5cd2962f1e8250620528c5b19075d34e uses `hasEffectiveAnnotation` instead of `hasAnnotation`. #164 has `? super T` and when the code asked whether the wildcard had a particular annotation in the lower bound, it got "none", because the type variable `T` isn't annotated. `hasEffectiveAnnotation` follows bounds to find the first declared type.

I replaced all `hasAnnotation` with `hasEffectiveAnnotation` in the `NullSpecAnnotatedTypeFactory`. I should probably have that done more incrementally, as two more conformance tests fail now... let me see why that is. I should also look at uses of `hasAnnotation` in other files.